Optimization of the affine transformations.

Sopel97 opened the pull request:
I pondered whether this should be a draft PR or an issue, but since I have code that can be compiled I decided a PR will be a better place for discussion. The code in this draft only supports target >= ssse3 so I removed others from the Makefile.

I've been tinkering lately with the manually vectorized affine transformations and I strongly believe that it can be easly improved. 
My idea is to unroll (for the hidden layers) the outer loop in https://github.com/official-stockfish/Stockfish/blob/0f6c08c73f516873b312cb8fce0d824a2167b075/src/nnue/layers/affine_transform.h#L112 by a factor of 4, accumulate to 4 separate sums, and then perform horizontal addition on each. It's faster than 4 separate horizontal additions (a lot of results are unused when doing hadd on one register at a time) and also allows to replace 4 32bit loads and and stores with one 128bit load and store.
SSSE3: https://godbolt.org/z/Kz1h5v
AVX2: https://godbolt.org/z/nqaoYK
AVX-512: https://godbolt.org/z/xM4E4T

I have measured a 4% speed increase for ssse3 build on my i7-920. I have also made similar adjustments to avx2, vnni256, avx512, vnni512 but I can't test them for speed because my CPU doesn't support the instruction sets. I've verified the correctness for avx2, vnni256, avx512, vnni512 using Intel Software Emulator (thanks @syzygy1, btw, may want to look at this).

I expect some discussion to take place here before this is tested on fishtest. Maybe some further improvements could be found. It would also be nice if someone benchmarked the >ssse3 versions.

Some random thoughts:
 - Maybe it would be better to do 8 parallel accumulators and haddx8 in AVX2 code, not sure about register pressure though
 - I tried to do 8 parallel accumulators for AVX512 but it lacks hadds and I couldn't figure how to do m512_haddx8 in a good way with shuffles + adds. AVX512 is fairly extensive so I expect some improvements and discoveries here.
 - For AVX512 it may worth to try larger sizes than 32 for the hidden layers.

-------------------------------

passed STC of the initial version:
LLR: 2.95 (-2.94,2.94) {-0.25,1.25}
Total: 17808 W: 1914 L: 1756 D: 14138
Ptnml(0-2): 76, 1330, 5948, 1460, 90
https://tests.stockfishchess.org/tests/view/5f9d516f6a2c112b60691da3

passed STC of the final version after cleanup:
LLR: 2.95 (-2.94,2.94) {-0.25,1.25}
Total: 16296 W: 1750 L: 1595 D: 12951
Ptnml(0-2): 72, 1192, 5479, 1319, 86
https://tests.stockfishchess.org/tests/view/5f9df5776a2c112b60691de3

Bench: 3870606

vondele commented on src/Makefile:
```diff
@@ -93,9 +93,7 @@ endif
-                 x86-64-sse41-popcnt x86-64-modern x86-64-ssse3 x86-64-sse3-popcnt \
-                 x86-64 x86-32-sse41-popcnt x86-32-sse2 x86-32 ppc-64 ppc-32 \
-                 armv7 armv7-neon armv8 apple-silicon general-64 general-32))
+                 x86-64-sse41-popcnt x86-64-modern x86-64-ssse3))
```
I guess that's part of the work in progress right?

Sopel97 replied:
Yes, I decided to remove the rest of the code for simplicity and only leave what matters here. For merge into master the code for <ssse3 would be included.

vondele commented:
On avx2, I measure about 1% speedup.
```
$ ~/chess/pyshbench/pyshbench.py ./stockfish.master ./stockfish.patch 10
run       base       test     diff
  1    2105092    2126002   +20910
  2    2110419    2144096   +33677
  3    2118367    2114663    -3704
  4    2092706    2118924   +26218
  5    2121526    2142196   +20670
  6    2093430    2134822   +41392
  7    2108027    2116699    +8672
  8    2209317    2228042   +18725
  9    2120596    2147904   +27308
 10    2197075    2223127   +26052

Result of  10 runs
==================
base (...kfish.master) =    2127656  +/- 25473
test (...ckfish.patch) =    2149648  +/- 25822
diff                   =     +21992  +/- 7793

speedup        = +0.0103
P(speedup > 0) =  1.0000

```

vondele commented:
as a minor nit, on avx2 some functions are unused:
```
nnue/../nnue/architectures/../layers/affine_transform.h:110:12: warning: variable ‘m256_hadd’ set but not used [-Wunused-but-set-variable]
  110 |       auto m256_hadd = [](__m256i sum, int bias) -> int {
      |            ^~~~~~~~~
nnue/../nnue/architectures/../layers/affine_transform.h:130:12: warning: variable ‘m128_hadd’ set but not used [-Wunused-but-set-variable]
  130 |       auto m128_hadd = [](__m128i sum, int bias) -> int {
      |            ^~~~~~~~~
nnue/../nnue/architectures/../layers/affine_transform.h:136:12: warning: variable ‘m128_haddx4’ set but not used [-Wunused-but-set-variable]
  136 |       auto m128_haddx4 = [](__m128i sum0, __m128i sum1, __m128i sum2, __m128i sum3, __m128i bias) -> __m128i {
```

Sopel97 commented:
I profiled the vnni512 version with the Intel Software Developement Emulator. If the results are to be believed they show that the most costly affine transform is almost twice as fast in the PR version.
Master: https://pastebin.com/UKLWjVMG in particular https://pastebin.com/UdyXvcaJ
PR: https://pastebin.com/adCYuC4X in particular https://pastebin.com/RgvsvtGV

danikyyzz commented on src/Makefile:
```diff
@@ -93,9 +93,7 @@ endif
 # the user can override with `make ARCH=x86-32-vnni256 SUPPORTED_ARCH=true`
 ifeq ($(ARCH), $(filter $(ARCH), \
                  x86-64-vnni512 x86-64-vnni256 x86-64-avx512 x86-64-bmi2 x86-64-avx2 \
-                 x86-64-sse41-popcnt x86-64-modern x86-64-ssse3 x86-64-sse3-popcnt \
-                 x86-64 x86-32-sse41-popcnt x86-32-sse2 x86-32 ppc-64 ppc-32 \
-                 armv7 armv7-neon armv8 apple-silicon general-64 general-32))
+                 x86-64-sse41-popcnt x86-64-modern x86-64-ssse3))
```
good

Sopel97 commented:
test here https://tests.stockfishchess.org/tests/view/5f9d516f6a2c112b60691da3

Sopel97 commented:
Is LTC required for speedup patches? If not I'll just fix the unused variables warnings and prepare a PR.

syzygy1 commented:
As far as I understand, LTC is not required for pure speed ups.

syzygy1 commented:
I think computing 8 inner-products in parallel should be possible on AVX2 without register spilling, but I haven't tried it yet.

How to do the horizontal adding most efficiently is an interesting question, especially on AVX512.
One possible approach is to first fold each 512-bit vector of partial sums into a 256-bit vector of partial sums by adding the upper part to the lower part. I don't know if that is faster than @Sopel97 's implementation.

It may now also make sense to clip the 32-bit sums immediately to 8-bit ints. This may be difficult to do in SF, though.

Sopel97 commented:
Sure, I'll document it and cover the cases that are currently `exit(1)`

syzygy1 commented on src/nnue/layers/affine_transform.h:
```diff
@@ -74,113 +74,578 @@ namespace Eval::NNUE::Layers {
-  #if !defined(USE_VNNI)
-      const __m256i kOnes = _mm256_set1_epi16(1);
-  #endif
+      // kOutputDimensions is either 1, 32, or 512
```
kOutputDimensions is either 1 or 32 (not 512)

syzygy1 commented on src/nnue/layers/affine_transform.h:
```diff
@@ -74,113 +74,578 @@ namespace Eval::NNUE::Layers {
+            for (IndexType j = 0; j < kNumChunks; ++j) {
+              const __m512i in = _mm512_loadA_si512(&input_vector512[j]);
+
+              sum0 = _mm512_dpbusd_epi32(sum0, in, _mm512_load_si512(&row0[j]));
```
I would leave out the explicit load instruction (so just `row0[j]`), and same for aligned stores. It looks cleaner, and it might give the compiler more opportunity for optimisation.

syzygy1 commented on src/nnue/layers/affine_transform.h:
```diff
@@ -74,113 +74,578 @@ namespace Eval::NNUE::Layers {
+          const auto row3 = reinterpret_cast<const __m256i*>(&weights_[offset3]);
+
+          IndexType j = 0;
+          for (; j < kNumChunks - 1; j += 2) {
```
It is safe to leave it to the compiler to unroll the loop.

Sopel97 replied:
yea, I'm tryout out snippets on godbolt now to see what can be simplified.

Sopel97 commented:
I tried 8x unroll at godbolt. With clang it looks very clean, with gcc not so much though and spills a few registers.
https://godbolt.org/z/hfdT9G

Just throwing this here for the future. I won't be changing this now.

Sopel97 commented:
I've made some changes to make the code cleaner and addressed the stuff syzygy pointed out. When it passes CI and gets @vondele 's approval I will put the final version on fishtest and then make a new PR.

Sopel97 commented:
I gave avx512 compiles for old and new versions to ipmanchess and he kindly ran both.
https://cdn.discordapp.com/attachments/771338720028393472/772192457290285086/Stockfish_avx512_no-pgo_old-new_benches.PNG
https://cdn.discordapp.com/attachments/771338720028393472/772192493885718548/Stockfish_avx512_no-pgo_benches.PNG

syzygy1 commented:
With Cfish, I found that the compiled code looks better if you replace the row0/1/2/3 pointers with a single pointer (looks like the difference between your clang and gcc output).

(I guess it's not so easy to look at the SF assembly code because SF relies heavily on LTO.)

syzygy1 commented:
Nice AVX512 speedup :)

Sopel97 commented:
I'm tinkering on godbolt. GCC seems to have `#pragma GCC unroll 4` which helps https://godbolt.org/z/41PsGa. Maybe such enforced unrolling could be forced further, though this is for another PR.

syzygy1 commented:
At least in Cfish the unrolling happens automatically after profiling.

It would be nice if one could make gcc produce assembly output even with LTO. I don't think it is possible, but I could easily be wrong. But you can still execute the g++ evaluate_nnue.cpp -S (with all the right flags) yourself and leave out -flto. You may have to prevent the Makefile from removing the .gcda files.

Sopel97 commented:
I think I'll look into using XED next time for such stuff (it comes with SDE). It seems to be east to find the assembly listings for the affine transforms.

vondele commented:
on linux you could do `CXXFLAGS="-g" make -j ARCH=x86-64-avx2 profile-build && objdump -dlS ./stockfish` to get the assembly, with source and line numbers intermixed.

I've done a few tests, seemingly a little faster now on avx2 as well. Compiled a couple of crosses etc. So looks good for a final test if you feel it is final. 

You can keep this PR, just force push a version with a full commit message including the testing results.

BTW, to test vnni etc, I have used cloud resources in the past, that's quite doable, but obviously no requirement. It is nice you've been able to check correctness on the emulator (please check again for the final version).

Sopel97 commented:
So a new net was merged and appveyor failed because it compares the bench to master. I've rebased and changed the bench in the commit message to match new master.
